akka-http-bench-jmh/src/main/scala/akka/http/impl/engine/http2/H2ServerProcessingBenchmark.scala

```diff
@@ -0,0 +1,95 @@
+import java.util.concurrent.CountDownLatch
+
+import scala.concurrent.Await
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import akka.actor.ActorSystem
+import akka.http.CommonBenchmark
+import akka.http.impl.engine.http2.FrameEvent.HeadersFrame
+import akka.http.impl.engine.http2.framing.FrameRenderer
+import akka.http.scaladsl.Http2
+import akka.http.scaladsl.model.HttpResponse
+import akka.http.scaladsl.settings.ServerSettings
+import akka.stream.{ ActorMaterializer, OverflowStrategy }
+import akka.stream.TLSProtocol.{ SslTlsInbound, SslTlsOutbound }
+import akka.stream.scaladsl.{ Flow, Keep, Sink, Source }
+import akka.util.ByteString
+import com.typesafe.config.ConfigFactory
+import org.openjdk.jmh.annotations._
+
+class H2ServerProcessingBenchmark extends CommonBenchmark {
+  // Obtained by converting the input request bytes from curl with --http2-prior-knowledge
+  def request(streamId: Int) =
+    FrameRenderer.render(HeadersFrame(streamId, endStream = true, endHeaders = true, HPackSpecExamples.C41FirstRequestWithHuffman, None))
+
+  val response: HttpResponse = HPackSpecExamples.FirstResponse
+
+  var httpFlow: Flow[ByteString, ByteString, Any] = _
+  implicit var system: ActorSystem = _
+  implicit var mat: ActorMaterializer = _
+
+  val packedResponse = ByteString(-62, -63, -64, -65, -66)
+
+  @Benchmark
+  @OperationsPerInvocation(10000)
+  def benchRequestProcessing(): Unit = {
+    val numRequests = 10000
+    val latch = new CountDownLatch(numRequests)
+
+    val (in, done) =
+      Source.queue(numRequests, OverflowStrategy.fail)
+        .viaMat(httpFlow)(Keep.left)
+        .toMat(Sink.foreach(res => {
+          // Skip headers/settings frames etc
+          if (res.containsSlice(HPackSpecExamples.C61FirstResponseWithHuffman)
+            || res.containsSlice(packedResponse)) {
+            latch.countDown()
+          }
+        }))(Keep.both)
+        .run()
+
+    in.offer(Http2Protocol.ClientConnectionPreface)
+
+    for (n <- Range(0, numRequests)) {
+      in.offer(request(1 + 2 * n))
+    }
+
+    latch.await()
+
+    in.complete()
+    Await.result(done, 1.hour)
+  }
+
+  @Setup
+  def setup(): Unit = {
+    val config =
+      ConfigFactory.parseString(
+        """
+           #akka.loglevel = debug
+           akka.actor.default-dispatcher.fork-join-executor.parallelism-max = 1
+           #akka.http.server.log-unencrypted-network-bytes = 100
+        """)
+        .withFallback(ConfigFactory.load())
+    system = ActorSystem("AkkaHttpBenchmarkSystem", config)
+    mat = ActorMaterializer()
+    val settings = implicitly[ServerSettings]
+    val log = system.log
+    implicit val ec = system.dispatcher
+    val http1 = Flow[SslTlsInbound].mapAsync(1)(_ => {
+      Future.failed[SslTlsOutbound](new IllegalStateException("Failed h2 detection"))
+    })
+    val http2 =
+      Http2Blueprint.handleWithStreamIdHeader(1)(req => {
+        req.entity.dataBytes.runWith(Sink.ignore).map(_ => response)
```

Comment:
```suggestion
        req.discardEntityBytes().future.map(_ => response)
```

which will avoid materializing an empty source.